### PR TITLE
Send CSM data only if consent is given

### PR DESCRIPTION
**Pull request by Amokrane**
The consent verification is made prior to sending metrics in the sending queue (in other words: what get puts into the sending queue, will be sent).

There are three states:
- CONSENT_GIVEN
- CONSENT_NOT_GIVEN
- CONSENT_UNKNOWN (when the response isn't retrieved from the backend yet).

When we are in the last state, we still write metrics. But, `CONSENT_GIVEN` is mandatory to send metrics to the sending queue.

### CODE REVIEW

**Comment by Amokrane** [comment]
WIP:
1. Need to fix the SONAR warnings.
2. In the scenario where consent is not given then given (before a metric is marked as `readyToSend`),  we will end up in a case where metric files will be half-written. I think that all CsmBidLifecycle callbacks, except the first one (CdbCallStarted), should not create a new file if it doesn't exist already.

File: publisher-sdk/src/main/java/com/criteo/publisher/model/CdbResponse.java [hunk 1]
```diff
@@ -76,7 +81,16 @@ public static CdbResponse fromJson(@NonNull JSONObject json) {
       }
     }
 
-    return new CdbResponse(slots, timeToNextCall);
+    boolean consentGiven = false;
+    if (json.has(CONSENT_GIVEN)) {
+      try {
+        consentGiven = json.getBoolean(CONSENT_GIVEN);
+      } catch (JSONException ex) {
+        logger.debug("Exception while reading consentGiven" + ex.getMessage());
```
Comment (TurpIF) [inline publisher-sdk/src/main/java/com/criteo/publisher/model/CdbResponse.java:89]: You should not swallow an exception. Please log the whole exception:
```
logger.debug("message", ex);
```

File: publisher-sdk/src/main/java/com/criteo/publisher/privacy/ConsentData.kt [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+   * <ul>
+   *  <li> CONSENT_GIVEN: it is known that consent has been given by the user</li>
+   *  <li> CONSENT_NOT_GIVEN: it is known that consent has not been given by the user</li>
+   *  <li> CONSENT_UNKNOWN: consent is not known yet (pending CDB request) </li>
+   * </ul>
+   */
+  enum class ConsentStatus {
+    CONSENT_GIVEN,
+    CONSENT_NOT_GIVEN,
+    CONSENT_UNKNOWN
```
Comment (TurpIF) [inline publisher-sdk/src/main/java/com/criteo/publisher/privacy/ConsentData.kt:37]: As told during the standup, we can drop the the unknown part and just keep a boolean.
The unknown correspond to no consent given.

File: publisher-sdk/src/main/java/com/criteo/publisher/model/CdbResponse.java [hunk 3]
```diff
@@ -76,7 +81,16 @@ public static CdbResponse fromJson(@NonNull JSONObject json) {
       }
     }
 
-    return new CdbResponse(slots, timeToNextCall);
+    boolean consentGiven = false;
```
Comment (TurpIF) [inline publisher-sdk/src/main/java/com/criteo/publisher/model/CdbResponse.java:84]: By having a default value to `false` instead of `null`, you should be sure that backend will always returns the consent even in case of no bid.

I'm ok with that, I made a ticket to check and implement this on backend: EE-1401

Comment (TurpIF) [inline publisher-sdk/src/main/java/com/criteo/publisher/model/CdbResponse.java:84]: By having a default value to `false` instead of `null`, you should be sure that backend will always returns the consent even in case of no bid.

I'm ok with that, I made a ticket to check and implement this on backend: EE-1401

File: publisher-sdk/src/main/java/com/criteo/publisher/model/CdbResponse.java (hunk 1 above)
Comment (TurpIF) [reply publisher-sdk/src/main/java/com/criteo/publisher/model/CdbResponse.java:89]: Btw, maybe could you apply this suggestion to the whole file ?

File: publisher-sdk/src/main/java/com/criteo/publisher/model/CdbResponse.java [hunk 4]
```diff
@@ -30,18 +30,23 @@
 
   private static final String TIME_TO_NEXT_CALL = "timeToNextCall";
   private static final String SLOTS = "slots";
+  private static final String CONSENT_GIVEN = "consentGiven";
```
Comment (TurpIF) [inline publisher-sdk/src/main/java/com/criteo/publisher/model/CdbResponse.java:33]: What do you think about transforming this class into an Java `@AutoValue` class or a Kotlin `data` class ?

Deserialization is isolating potential error per slots, but except the prefetch (which might be removed), all bid response contains only a single slot.

File: test-utils/src/main/java/com/criteo/publisher/TestDeviceIds.kt [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+package com.criteo.publisher
+
+const val NON_CONSENTED_DEVICE_ID = "2"
```
Comment (TurpIF) [inline test-utils/src/main/java/com/criteo/publisher/TestDeviceIds.kt:19]: Nit: the project already contains the `StubConstants` class containing constants.
I think that this file will not contains anything but this line. What do you think about moving this into `StubConstants` instead ?

**Review by TurpIF — CHANGES_REQUESTED** [review]
As told during the stand up, functional change is requested:
- SDK only treat consent as a boolean value (given or not)
- By default, consent is not given
- Consent is cached in shared pref
- For each CDB requests, consent is updated (this implies an accepted lag of 1 bid)
- If consent is not given, then CSM and remote log are disabled
- (1) To determine: in case of transition from consent given to consent not given, what should we do on persisted CSM/logs:
  - Clean persisted data: more work
  - Do nothing: consent was given when gathering such data
- (2) To determine: in case of transition from consent not given to consent given, should we accept partially filled CSM metrics ?
  - Yes: no work to do, but partial metrics
  - No: add a lot of complexity, for always filled metrics

On (1) I would tend to do nothing: it is edge case adding complexity.
On (2) I would tend to accept partial metrics. Partial metrics can already occurs (in case of thrown exception for instance). The transition is an edge case so expected volume of such event is low. We can react later if we see issues on CSM.

File: test-utils/src/main/java/com/criteo/publisher/TestDeviceIds.kt (hunk 5 above)
Comment (Amokrane) [reply test-utils/src/main/java/com/criteo/publisher/TestDeviceIds.kt:19]: Well, StubConstants contains .. stub constants which isn't the case of this deviceId. I feel that even if there is only a single constant in this file, it is still worth separating concerns and having clear boundaries. Or we should rename `StubConstants` into `TestConstants` or something like that.

File: publisher-sdk/src/main/java/com/criteo/publisher/model/CdbResponse.java (hunk 4 above)
Comment (Amokrane) [reply publisher-sdk/src/main/java/com/criteo/publisher/model/CdbResponse.java:33]: I think this is a good idea, but I don't think we should do this in this PR. It should be done separately.

File: test-utils/src/main/java/com/criteo/publisher/TestDeviceIds.kt (hunk 5 above)
Comment (TurpIF) [reply test-utils/src/main/java/com/criteo/publisher/TestDeviceIds.kt:19]: Renaming `StubConstants` will be a breaking change in lot of tests. Included ones of mediation adapters.
Ok to keep this file then. Maybe should you encapsulate the constant in an kotlin `object` or use `@file:JvmName` to make it nicer from Java code (see https://stackoverflow.com/a/57355475/3663982 to make your choice).

File: publisher-sdk/src/main/java/com/criteo/publisher/privacy/ConsentData.kt (hunk 2 above)
Comment (Amokrane) [reply publisher-sdk/src/main/java/com/criteo/publisher/privacy/ConsentData.kt:37]: I am ok with dropping the `CONSENT_UNKNOWN` constant. Note that we are loosing the ability to distinguish between `CONSENT_NOW_GIVEN` and `CONSENT_UNKNOWN` in console logs, in this case.

File: publisher-sdk/src/main/java/com/criteo/publisher/model/CdbResponse.java (hunk 1 above)
Comment (Amokrane) [reply publisher-sdk/src/main/java/com/criteo/publisher/model/CdbResponse.java:89]: Yeah was gonna say that I just copy/pasted whatever was done above :) (not that it's a good excuse).

File: publisher-sdk/src/main/java/com/criteo/publisher/privacy/ConsentData.kt [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  }
+  private val valueRef = AtomicReference<ConsentStatus>(ConsentStatus.CONSENT_UNKNOWN)
+
+  fun getConsentStatus(): ConsentStatus = valueRef.get()
```
Comment (TurpIF) [inline publisher-sdk/src/main/java/com/criteo/publisher/privacy/ConsentData.kt:41]: Nit: this can be a kotlin `var` with `get()/set()`

**Comment by Amokrane** [comment]
> As told during the stand up, functional change is requested:
> 
> * SDK only treat consent as a boolean value (given or not)
> * By default, consent is not given
> * Consent is cached in shared pref
> * For each CDB requests, consent is updated (this implies an accepted lag of 1 bid)
> * If consent is not given, then CSM and remote log are disabled
> * (1) To determine: in case of transition from consent given to consent not given, what should we do on persisted CSM/logs:
>   
>   * Clean persisted data: more work
>   * Do nothing: consent was given when gathering such data
> * (2) To determine: in case of transition from consent not given to consent given, should we accept partially filled CSM metrics ?
>   
>   * Yes: no work to do, but partial metrics
>   * No: add a lot of complexity, for always filled metrics
> 
> On (1) I would tend to do nothing: it is edge case adding complexity.
> On (2) I would tend to accept partial metrics. Partial metrics can already occurs (in case of thrown exception for instance). The transition is an edge case so expected volume of such event is low. We can react later if we see issues on CSM.


> To determine: in case of transition from consent given to consent not given, what should we do on persisted CSM/logs:

For that transition, we also need to make sure that we don't further write any CSM/logs after the transition. If we don't do anything, then we will continue writing new CSMs/logs during the init phase of the SDK (and until we receive a response from Cdb). I think we should monitor if any of the CMP values have changed on the client side and change the consent back to `CONSENT_NOT_GIVEN` in that case. This can be done either during init only (by keeping an old copy of the CMP data) or by registering callbacks. WDYT?

File: publisher-sdk/src/androidTest/java/com/criteo/publisher/csm/CsmFunctionalTest.java [hunk 7]
```diff
@@ -59,14 +66,28 @@
 import org.junit.Before;
 import org.junit.Rule;
 import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameter;
 import org.mockito.MockitoAnnotations;
 
 @FlakyTest(detail = "CSM are triggered concurrently of bid")
+@RunWith(Parameterized.class)
```
[the remark this answers is not in the document]
Comment (Amokrane) [reply publisher-sdk/src/androidTest/java/com/criteo/publisher/csm/CsmFunctionalTest.java:75]: Good to know, I'll check.

**Comment by TurpIF** [comment]
> > * (1) To determine: in case of transition from consent given to consent not given, what should we do on persisted CSM/logs:
> >   
> >   * Clean persisted data: more work
> >   * Do nothing: consent was given when gathering such data
> > * (2) To determine: in case of transition from consent not given to consent given, should we accept partially filled CSM metrics ?
> >   
> >   * Yes: no work to do, but partial metrics
> >   * No: add a lot of complexity, for always filled metrics
> > 
> > On (1) I would tend to do nothing: it is edge case adding complexity.
> > On (2) I would tend to accept partial metrics. Partial metrics can already occurs (in case of thrown exception for instance). The transition is an edge case so expected volume of such event is low. We can react later if we see issues on CSM.
> 
> Ok for both transitions.
> 
> However, for:
> 
> > To determine: in case of transition from consent given to consent not given, what should we do on persisted CSM/logs:
> 
> For that transition, we also need to make sure that we don't further write any CSM/logs after the transition. If we don't do anything, then we will continue writing new CSMs/logs during the init phase of the SDK (and until we receive a response from Cdb). I think we should monitor if any of the CMP values have changed on the client side and change the consent back to `CONSENT_NOT_GIVEN` in that case. This can be done either during init only (by keeping an old copy of the CMP data) or by registering callbacks. WDYT?

By transition, I meant transition in backend response: so SDK is just sending the consent strings to the backend and listening the response. So the consent strings are never read by the SDK.

There you're talking about transition in the consent strings. Which is 1 bid ahead. But I thought that we accepted this delay:

> For each CDB requests, consent is updated (this implies an accepted lag of 1 bid)

Please note that the delay is symmetrical. It is not to bias the decision in any way. This is to truly reflect the backend decision.

Moreover, keeping a state of consent strings (GDPR, CCPA, MoPub, iOS, ...) and invalidating consent based on this can:
- at best: do its job with the cost of technical complexity
- at worst: kill CSM or remote logging features because consent strings may legitimately change: for instance if they contains a timestamp, ...)

Adding the fact that transitions are uncommon edge cases (I really never saw any app asking twice a consent), I would prefer to only handle transitions in backend response, not transitions in consent strings.

**Comment by Amokrane** [comment]
> By transition, I meant transition in backend response: so SDK is just sending the consent strings to the backend and listening the response. So the consent strings are never read by the SDK.

I am not sure I get this point: `So the consent strings are never read by the SDK.`. It is the SDK that reads the consent strings, not the backend. reading != interpreting or treating.

> Adding the fact that transitions are uncommon edge cases (I really never saw any app asking twice a consent), I would prefer 
to only handle transitions in backend response, not transitions in consent strings.

It's not that the app will ask the consent twice, but users can and (will) modify their consent by going to the device or app settings. 
See the screen below, taken from LeMonde application, where I can modify my consent.
Webp net-resizeimage (1)

If I remove my consent when will the app take into account the new consent and stop sending CSMs/logs? With the current implementation, it will happen after the next Cdb response and during the time window before that next Cdb response all my data will still be logged which does not really comply with legislation I suppose :)

> There you're talking about transition in the consent strings. Which is 1 bid ahead. But I thought that we accepted this delay:

It is acceptable if we are missing logs in the case where we don't yet that consent was given. Whether it is  acceptable to log data that we shouldn't log, (even for a 1-bid lag) shouldn't be decided by us IMHO, let's double check with PM on this.

> at best: do its job with the cost of technical complexity
at worst: kill CSM or remote logging features because consent strings may legitimately change: for instance if they contains a timestamp, ...)

Consent data don't have any timestamp and won't change unless the user has changed them, or unless there is a new TCF/CCPA version.  Technical complexity: well it's not that complex to listen for SharedPref changes or keep a copy of the old data.

I am ok to go with the simple solution, but we should get an explicit agreement over this from PM.

**Comment by TurpIF** [comment]
> > By transition, I meant transition in backend response: so SDK is just sending the consent strings to the backend and listening the response. So the consent strings are never read by the SDK.
> 
> I am not sure I get this point: `So the consent strings are never read by the SDK.`. It is the SDK that reads the consent strings, not the backend. reading != interpreting or treating.

Yep sorry, I wrongly express myself. Instead of `read`, I meant `used` (except for forging the bid request).
 
> > Adding the fact that transitions are uncommon edge cases (I really never saw any app asking twice a consent), I would prefer
> > to only handle transitions in backend response, not transitions in consent strings.
> 
> It's not that the app will ask the consent twice, but users can (and will) modify their consent by going to the device or app settings.
> See the screen below, taken from LeMonde application, where I can modify my consent.
> Webp net-resizeimage (1)

To be honest, that's the first time I see a screen like that.

> I am ok to go with the simple solution, but we should get an explicit agreement over this from PM.

Ok, let's ask to him next week. Meanwhile, I think that setting up listener+consent invalidation can be done on top of this PR. What do you think to move on and add a new ticket as a reminder ?

**Comment by Amokrane** [comment]
> Ok, let's ask to him next week. Meanwhile, I think that setting up listener+consent invalidation can be done on top of this PR. What do you think to move on and add a new ticket as a reminder ?

Sounds good to me, let's do that.

File: publisher-sdk/src/androidTest/java/com/criteo/publisher/csm/CsmFunctionalTest.java [hunk 8]
```diff
@@ -59,14 +66,28 @@
...
+  @Parameterized.Parameters
+  public static Collection consents() {
+    return Arrays.asList(new Object[][] {
+        { true },
+        { false }
+    });
+  }
+
+  @Parameter(0)
+  public boolean consentGiven;
```
[the remark this answers is not in the document]
Comment (Amokrane) [reply publisher-sdk/src/androidTest/java/com/criteo/publisher/csm/CsmFunctionalTest.java]: After working again on the change I am now thinking that these functional tests should only cover the nominal case: when CSM is enabled. The case where CSM is disabled (either due to consent or CaC which by the way isn't covered here) is already covered in unit tests.

File: publisher-sdk/src/main/java/com/criteo/publisher/privacy/ConsentData.kt (hunk 6 above)
Comment (Amokrane) [reply publisher-sdk/src/main/java/com/criteo/publisher/privacy/ConsentData.kt:40]: If I do that then I would need to expose the `AtomicReference`. I am finding it cleaner to expose the value directly. If I don't use an `AtomicReference`, then I would need to make it volatile. (the consent status is written/read on different threads within code that is not synchronized).

Comment (TurpIF) [reply publisher-sdk/src/main/java/com/criteo/publisher/privacy/ConsentData.kt:40]: The `valueRef` can stay `private` and a `var` could use it:
```
private val valueRef = atomic

var consentStatus: ConsentStatus
  get() = valueRef.get()
  set(value) = valueRef.set(value)
```

File: publisher-sdk/src/main/java/com/criteo/publisher/privacy/ConsentData.kt [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+package com.criteo.publisher.privacy
+
+import com.criteo.publisher.annotation.OpenForTesting
+import java.util.concurrent.atomic.AtomicReference
+
+@OpenForTesting
+class ConsentData {
+
+  /**
+   * Consent can be in any of the following three states:
```
Comment (TurpIF) [inline publisher-sdk/src/main/java/com/criteo/publisher/privacy/ConsentData.kt:26]: `s/three/two/`

File: publisher-sdk/src/main/java/com/criteo/publisher/privacy/ConsentData.kt [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+   *
+   * <ul>
+   *  <li> CONSENT_GIVEN: it is known that consent has been given by the user</li>
+   *  <li> CONSENT_NOT_GIVEN: it is known that consent has not been given by the user,
+   *  or isn't known yet.</li>
+   * </ul>
+   */
+  enum class ConsentStatus {
```
Comment (TurpIF) [inline publisher-sdk/src/main/java/com/criteo/publisher/privacy/ConsentData.kt:34]: Do you have a reason to use an enum value rather than a simple boolean ?
Using enum makes implementation a little bit harder to get.

File: publisher-sdk/src/main/java/com/criteo/publisher/privacy/ConsentData.kt [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+    CONSENT_GIVEN,
+    CONSENT_NOT_GIVEN
+  }
+  private val valueRef = AtomicReference<ConsentStatus>(ConsentStatus.CONSENT_NOT_GIVEN)
```
Comment (TurpIF) [inline publisher-sdk/src/main/java/com/criteo/publisher/privacy/ConsentData.kt:38]: This value is not persisted. So you're always losing the consent at each new session. 
Do you plan to persist it in another PR ?

File: publisher-sdk/src/test/java/com/criteo/publisher/LiveCdbCallListenerTests.kt [hunk 12]
```diff
@@ -182,4 +186,22 @@ class LiveCdbCallListenerTests {
     verify(bidLifecycleListener, never()).onBidConsumed(any(), any())
     verify(bidLifecycleListener).onCdbCallFailed(cdbRequest, exception)
   }
+
+  @Test
+  fun onBidResponse_givenConsentGiven_ThenUpdatConsentDataAccordingly() {
```
Comment (TurpIF) [inline publisher-sdk/src/test/java/com/criteo/publisher/LiveCdbCallListenerTests.kt:191]: s/Updat/Update

File: publisher-sdk/src/test/java/com/criteo/publisher/LiveCdbCallListenerTests.kt [hunk 13]
```diff
@@ -182,4 +186,22 @@ class LiveCdbCallListenerTests {
...
+    whenever(cdbResponse.consentGiven).thenReturn(true)
+
+    liveCdbCallListener.onCdbResponse(cdbRequest, cdbResponse)
+
+    verify(consentData).consentStatus = ConsentData.ConsentStatus.CONSENT_GIVEN
+  }
+
+  @Test
+  fun onBidResponse_givenConsentNotGiven_ThenUpdatConsentDataAccordingly() {
```
Comment (TurpIF) [inline publisher-sdk/src/test/java/com/criteo/publisher/LiveCdbCallListenerTests.kt:200]: s/Updat/Update

File: publisher-sdk/src/test/java/com/criteo/publisher/csm/CsmBidLifecycleListenerTest.kt [hunk 14]
```diff
@@ -96,10 +104,12 @@ class CsmBidLifecycleListenerTest {
   }
 
   @Test
-  fun onSdkInitialized_PushAllMetricsInQueue() {
+  fun onSdkInitialized_GivenConsentNotGiven_DoNothing() {
```
Comment (TurpIF) [inline publisher-sdk/src/test/java/com/criteo/publisher/csm/CsmBidLifecycleListenerTest.kt:107]: There is no more test covering the normal case: on SDK init, CSM are pushed in the queue

File: publisher-sdk/src/test/java/com/criteo/publisher/csm/CsmBidLifecycleListenerTest.kt [hunk 15]
```diff
@@ -284,8 +257,20 @@ class CsmBidLifecycleListenerTest {
...
+
+    listener.onBidConsumed(adUnit, mock())
+
+    verifyFeatureIsDeactivated()
+  }
+
+
   @Test
   fun onBidConsumed_GivenNotExpiredBid_SetElapsedTimeAndReadyToSend() {
+    givenConsentGiven()
```
Comment (TurpIF) [inline publisher-sdk/src/test/java/com/criteo/publisher/csm/CsmBidLifecycleListenerTest.kt:273]: Nit: not useful as your setup already gives the consent

File: publisher-sdk/src/test/java/com/criteo/publisher/csm/CsmBidLifecycleListenerTest.kt [hunk 16]
```diff
@@ -303,12 +288,11 @@ class CsmBidLifecycleListenerTest {
       verify(it).setElapsedTimestamp(42)
       verify(it).setReadyToSend(true)
     }
-
-    verify(sendingQueueProducer).pushInQueue(repository, "id")
   }
 
   @Test
   fun onBidConsumed_GivenExpiredBid_SetReadyToSend() {
+    givenConsentGiven()
```
Comment (TurpIF) [inline publisher-sdk/src/test/java/com/criteo/publisher/csm/CsmBidLifecycleListenerTest.kt:295]: Nit: not useful as your setup already gives the consent

File: publisher-sdk/src/test/java/com/criteo/publisher/csm/CsmBidLifecycleListenerTest.kt [hunk 17]
```diff
@@ -141,54 +161,8 @@ class CsmBidLifecycleListenerTest {
...
-
-    listener.onCdbCallFinished(request, response)
-
-    verify(sendingQueueProducer, never()).pushInQueue(any(), any())
-  }
-
-  @Test
-  fun onCdbCallFinished_GivenNoBidAndInvalidBidAndValidBidReceived_UpdateThemByIdAccordingly() {
+  fun onCdbCallFinished_GivenConsentGivenAndNoBidAndInvalidBidAndValidBidReceived_UpdateThemByIdAccordingly() {
+    givenConsentGiven()
```
Comment (TurpIF) [inline publisher-sdk/src/test/java/com/criteo/publisher/csm/CsmBidLifecycleListenerTest.kt:165]: Nit: not useful as your setup already gives the consent

File: publisher-sdk/src/test/java/com/criteo/publisher/csm/CsmBidLifecycleListenerTest.kt [hunk 18]
```diff
@@ -132,7 +151,8 @@ class CsmBidLifecycleListenerTest {
   }
 
   @Test
-  fun onCdbCallFinished_GivenDeactivatedFeature_DoNothing() {
+  fun onCdbCallFinished_GivenConsentGivenAndDeactivatedFeature_DoNothing() {
+    givenConsentGiven()
```
Comment (TurpIF) [inline publisher-sdk/src/test/java/com/criteo/publisher/csm/CsmBidLifecycleListenerTest.kt:155]: Nit: not useful as your setup already gives the consent

File: publisher-sdk/src/test/java/com/criteo/publisher/csm/CsmBidLifecycleListenerTest.kt [hunk 19]
```diff
@@ -338,6 +320,24 @@ class CsmBidLifecycleListenerTest {
     verifyZeroInteractions(repository)
   }
 
+  @Test
+  fun onBidsCached_GivenDeactivatedFeature_DoNothing() {
+   givenDeactivatedFeature()
```
Comment (TurpIF) [inline publisher-sdk/src/test/java/com/criteo/publisher/csm/CsmBidLifecycleListenerTest.kt:325]: There is a missing space here

File: publisher-sdk/src/test/java/com/criteo/publisher/csm/CsmBidLifecycleListenerTest.kt [hunk 20]
```diff
@@ -141,54 +161,8 @@ class CsmBidLifecycleListenerTest {
   }
 
   @Test
-  fun onCdbCallFinished_GivenOnlyNoBid_PushReadyToSendInQueue() {
```
Comment (TurpIF) [inline publisher-sdk/src/test/java/com/criteo/publisher/csm/CsmBidLifecycleListenerTest.kt:144]: Why are all those tests/assertions removed ?

File: test-utils/src/main/java/com/criteo/publisher/network/CdbMock.kt [hunk 21]
```diff
@@ -141,9 +141,11 @@ class CdbMock(private val jsonSerializer: JsonSerializer) {
 
       val responseSlots = cdbRequest.slots.mapNotNull { it.toResponseSlot() }.joinToString()
       val requestId = UUID.randomUUID().toString()
+
       val cdbResponse = """
-      {
+      {z
```
Comment (TurpIF) [inline test-utils/src/main/java/com/criteo/publisher/network/CdbMock.kt:146]: There is an extra `z`

**Review by TurpIF — CHANGES_REQUESTED** [review]

File: publisher-sdk/src/androidTest/java/com/criteo/publisher/csm/CsmFunctionalTest.java [hunk 22]
```diff
@@ -404,4 +430,11 @@ private void waitForIdleState() {
   private void loadBid(@NonNull AdUnit adUnit) {
     Criteo.getInstance().loadBid(adUnit, new ContextData(), ignored -> { /* no op */ });
   }
+
+  private void triggerMetricRequest() {
+    threadPoolExecutor.execute(() -> metricSendingQueueProducer.pushAllInQueue(metricRepository));
```
[the remark this answers is not in the document]
Comment (Amokrane) [reply publisher-sdk/src/androidTest/java/com/criteo/publisher/csm/CsmFunctionalTest.java:435]: Ah I didn't know about `MetricHelper`, will do.

File: publisher-sdk/src/main/java/com/criteo/publisher/csm/CsmBidLifecycleListener.java [hunk 23]
```diff
@@ -287,6 +290,6 @@ private void updateByCdbRequestIds(@NonNull CdbRequest request, @NonNull MetricU
   }
 
   private boolean isCsmDisabled() {
-    return !config.isCsmEnabled();
+    return !config.isCsmEnabled() || consentData.getConsentStatus().equals(CONSENT_NOT_GIVEN);
```
[the remark this answers is not in the document]
Comment (Amokrane) [reply publisher-sdk/src/main/java/com/criteo/publisher/csm/CsmBidLifecycleListener.java:293]: Yep sure.

File: publisher-sdk/src/main/java/com/criteo/publisher/csm/CsmBidLifecycleListener.java [hunk 24]
```diff
@@ -58,21 +65,17 @@ public CsmBidLifecycleListener(
...
   ) {
     this.repository = repository;
     this.sendingQueueProducer = sendingQueueProducer;
     this.clock = clock;
     this.config = config;
+    this.consentData = consentData;
     this.executor = executor;
   }
 
-  /**
```
[the remark this answers is not in the document]
Comment (Amokrane) [reply publisher-sdk/src/main/java/com/criteo/publisher/csm/CsmBidLifecycleListener.java:70]: Removed it when I wrote the first implementation where onSdkinit was no longer doing anything. Putting the comment back.

File: publisher-sdk/src/main/java/com/criteo/publisher/csm/MetricDirectory.java [hunk 25]
```diff
@@ -105,7 +106,6 @@ private String getMetricFilenameFromImpressionId(@NonNull String impressionId) {
    * This is the dual of {@link #getMetricFilenameFromImpressionId(String)}: The impression ID is a
    * string, and the metric filename is the impression ID with an extension. To find again the
    * impression ID from the filename, then this methods only have to remove the extension.
-   *
```
[the remark this answers is not in the document]
Comment (Amokrane) [reply publisher-sdk/src/main/java/com/criteo/publisher/csm/MetricDirectory.java:108]: Yeah, actually I added code with the previous implementation which I removed back again.

File: publisher-sdk/src/main/java/com/criteo/publisher/privacy/ConsentData.kt (hunk 10 above)
Comment (Amokrane) [reply publisher-sdk/src/main/java/com/criteo/publisher/privacy/ConsentData.kt:34]: I feel it is cleaner and makes the code easier to read. It's unclear seeing `true/false` being passed around vs. seeing values such as these ones.

File: publisher-sdk/src/main/java/com/criteo/publisher/privacy/ConsentData.kt (hunk 11 above)
Comment (Amokrane) [reply publisher-sdk/src/main/java/com/criteo/publisher/privacy/ConsentData.kt:38]: Yep planned for another PR.

File: publisher-sdk/src/test/java/com/criteo/publisher/csm/CsmBidLifecycleListenerTest.kt (hunk 14 above)
Comment (Amokrane) [reply publisher-sdk/src/test/java/com/criteo/publisher/csm/CsmBidLifecycleListenerTest.kt:107]: Good catch, I removed it back when I wrote the first implementation.

File: publisher-sdk/src/androidTest/java/com/criteo/publisher/csm/CsmFunctionalTest.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (Amokrane) [reply publisher-sdk/src/androidTest/java/com/criteo/publisher/csm/CsmFunctionalTest.java]: Though not gonna lie, it is an `enum` because initially there were three states. I can change it to `boolean` in the next PR where the consent is stored.

File: publisher-sdk/src/main/java/com/criteo/publisher/privacy/ConsentData.kt (hunk 10 above)
Comment (TurpIF) [reply publisher-sdk/src/main/java/com/criteo/publisher/privacy/ConsentData.kt:34]: It might be cleaner internally to the `ConsentData` class. Although, even if there was a third state, I would argue that callers should not be able to know about internal state and should only consume a boolean decision: do I have the consent or not ?

Comment (Amokrane) [reply publisher-sdk/src/main/java/com/criteo/publisher/privacy/ConsentData.kt:34]: Yeah I will move it to boolean in another PR.

**Comment by Amokrane** [comment]
I am fixing the code smells.

**Comment by Amokrane** [comment]
Actually, to minimize build and waiting times, I would rather fix the minor code smells on the other PR.

**Review by TurpIF — APPROVED** [review]
Ok for doing the minor fix in another PR
